src/rust_api.rs

```diff
@@ -0,0 +1,551 @@
+
+        Ok(())
+    }
+
+    /// Get the next decoded frame from the decoder.
+    ///
+    /// If this returns `Err([Error::Again])` then further data has to be sent to the decoder
+    /// before further decoded frames become available.
+    ///
+    /// To make most use of frame threading this function should only be called once per submitted
+    /// input frame and not until it returns `Err([Error::Again])`. Calling it in a loop should
+    /// only be done to drain all pending frames at the end.
+    pub fn get_picture(&mut self) -> Result<Picture, Rav1dError> {
+        let mut pic = Rav1dPicture::default();
+        let ret = rav1d_get_picture(&self.ctx, &mut pic);
+
+        if let Err(err) = ret {
+            Err(err)
+        } else {
+            Ok(Picture {
+                inner: Arc::new(pic),
+            })
+        }
+    }
+
+    /// Get the decoder delay.
+    pub fn get_frame_delay(&self) -> u32 {
+        // The only fields this actually needs from Rav1dSettings are n_threads and max_frame_delay so we just pass these in directly
+
+        rav1d_get_frame_delay(&Rav1dSettings {
+            n_threads: self.n_threads,
+            max_frame_delay: self.max_frame_delay,
+            ..Default::default()
+        }) as u32
+    }
+}
+
+impl Drop for Decoder {
+    fn drop(&mut self) {
+        if let Some(mut pending_data) = self.pending_data.take() {
+            let _ = mem::take(&mut pending_data);
+        }
+        rav1d_close(self.ctx.clone());
+    }
+}
+
+static_assertions::assert_impl_all!(Decoder: Send, Sync);
+
+/// A decoded frame.
+#[derive(Clone)]
+pub struct Picture {
+    inner: Arc<Rav1dPicture>,
+}
+
+/// Frame component.
+#[derive(Eq, PartialEq, Copy, Clone, Debug)]
+pub enum PlanarImageComponent {
+    /// Y component (Luminance).
+    Y,
+    /// U component (Chrominance).
+    U,
+    /// V component (Chrominance).
+    V,
+}
+
+impl TryFrom<usize> for PlanarImageComponent {
+    type Error = Rav1dError;
+    fn try_from(index: usize) -> Result<Self, Self::Error> {
+        match index {
+            0 => Ok(PlanarImageComponent::Y),
+            1 => Ok(PlanarImageComponent::U),
+            2 => Ok(PlanarImageComponent::V),
+            _ => Err(Rav1dError::InvalidArgument),
+        }
+    }
+}
+
+impl From<PlanarImageComponent> for usize {
+    fn from(component: PlanarImageComponent) -> Self {
+        match component {
+            PlanarImageComponent::Y => 0,
+            PlanarImageComponent::U => 1,
+            PlanarImageComponent::V => 2,
+        }
+    }
+}
+
+/// A single plane of a decoded frame.
+///
+/// This can be used like a `&[u8]`.
+#[derive(Clone)]
+pub struct Plane(Picture, PlanarImageComponent);
+
+impl AsRef<[u8]> for Plane {
+    fn as_ref(&self) -> &[u8] {
+        let (stride, height) = self.0.plane_data_geometry(self.1);
+        // SAFETY: both stride and height can't be negative, the `stride` and `height` methods panic if they are so there's no undefined behaviour
```

Comment:
The wording here misses the mark and ends in a non-sequitur. It should justify why the extent is correct for the pointer that's used but only justifies how the inputs are somehow bounded in `[0; u32::MAX)` (repeating the type explicitly would also help readability).